naga/src/front/atomic_upgrade.rs

```diff
@@ -0,0 +1,238 @@
+        let padding = self.inc_padding();
+        padding.trace("upgrading type: ", ty);
+
+        let r#type = self.module.types[ty].clone();
+
+        let inner = match r#type.inner.clone() {
+            TypeInner::Scalar(scalar) => {
+                log::trace!("{padding}hit the scalar leaf, replacing with an atomic");
+                TypeInner::Atomic(scalar)
+            }
+            TypeInner::Pointer { base, space } => TypeInner::Pointer {
+                base: self.upgrade_type(base)?,
+                space,
+            },
+            TypeInner::Array { base, size, stride } => TypeInner::Array {
+                base: self.upgrade_type(base)?,
+                size,
+                stride,
+            },
+            TypeInner::Struct { members, span } => TypeInner::Struct {
+                members: {
+                    let mut new_members = vec![];
+                    for member in members.iter().cloned() {
+                        let StructMember {
+                            name,
+                            ty,
+                            binding,
+                            offset,
+                        } = member;
+                        new_members.push(StructMember {
+                            name,
+                            ty: self.upgrade_type(ty)?,
+                            binding,
+                            offset,
+                        });
+                    }
+                    new_members
+                },
+                span,
+            },
+            TypeInner::BindingArray { base, size } => TypeInner::BindingArray {
+                base: self.upgrade_type(base)?,
+                size,
+            },
+            n => n,
+        };
+
+        let new_type = Type {
+            name: r#type.name.clone(),
+            inner,
+        };
+        let new_ty = if let Some(prev_ty) = self.module.types.get(&new_type) {
+            padding.trace("type exists: ", prev_ty);
+            prev_ty
+        } else {
+            padding.debug("ty: ", ty);
+            padding.debug("from: ", &r#type);
+            padding.debug("to:   ", &new_type);
+
+            let new_ty = self
+                .module
+                .types
+                .insert(new_type, self.module.types.get_span(ty));
+            padding.debug("new ty: ", new_ty);
+            new_ty
+        };
+        Ok(new_ty)
+    }
+
+    fn upgrade_global_variable(&mut self, handle: Handle<GlobalVariable>) -> Result<(), Error> {
+        let padding = self.inc_padding();
+        padding.trace("upgrading global variable: ", handle);
+
+        let var = self.module.global_variables[handle].clone();
+
+        let new_var = GlobalVariable {
+            name: var.name.clone(),
+            space: var.space,
+            binding: var.binding.clone(),
+            ty: self.upgrade_type(var.ty)?,
+            init: self.upgrade_opt_expression(None, var.init)?,
```

Comment:
If we don't need to upgrade the global variable's `init` here, then I think we can get rid of a lot more code. It would remove the need for `upgrade_expression` and `upgrade_opt_expression`.

Reply:
That could be. Take a look at what the `valid` module does with `GlobalVariable::init`. You should be able to tell whether it will treat an `atomic<T>` variable like a `T` variable.

Reply:
(I often resort to reading `naga/src/valid` when I want to understand the expectations of Naga IR. Ideally each time I did some exploration I would add docs to `lib.rs`...)

Reply:
From chat:

> It seems like validate_global_var, at least, does not allow atomic globals to have initializers at all. Initializers are only allowed in the private and function address spaces, and atomics are not allowed in either of those.
>
> It seems that WGSL also only permits initializers for variables in the private or function address spaces, so that bit of Naga is faithfully enforcing WGSL's rules.
>
> For the sake of incremental development, it would probably make sense to just forbid initializers for now.